example/main.go

```diff
@@ -21,17 +23,21 @@ func main() {
 		flag.Usage()
 		return
 	}
-	db, err := notmuch.Open(*dir, notmuch.DB_RDONLY)
+	db, err := notmuch.Open(*dir, notmuch.DBReadOnly)
 	if err != nil {
 		fmt.Println(err)
 		return
 	}
 	defer db.Close()
-	threads, err := db.QueryCreate(*queryString).SearchThreads()
+	threads, err := db.NewQuery(*queryString).Threads()
 	if err != nil {
 		fmt.Println(err)
 	}
-	for ; threads.Valid(); threads.MoveToNext() {
+	for {
+		if err := threads.MoveToNext(); err != nil {
+			break
+		}
+
```

Comment:
These are semantically different; The old version does MoveToNext at the *end* of the loop, yours does so at the beginning. As such, you have a bug; you're skipping the first result. The query `aur-general` returns one result on the old version, none on the new one.

You've also removed `Valid`, so it's impossible to implement the example correctly; If you do a `Get()` on an empty result it will panic. We can discuss what the idiom should be, but there needs to be a way of handling this. Ideas:

1. Have `Get()` first call valid under the covers, and return the error instead of (or in addition to) MoveToNext().
2. Keep exposing `Valid`.

I'm leaning towards having both methods return an error; the interfaces we have now (both the old version and this one) come to close to panics being a part of the interface for me to be happy. We should probably still implement a `valid` wrapper. I'm on the fence as to whether I think it should be exported, but it's easier to add than remove so probably should have it be private for now.